src/out-of-bounds-access.c

```diff
@@ -3,56 +3,64 @@
 #include <string.h>
 #include <stdlib.h>
 
-// An example of a user changing the root password by entering a username that's long enough (>12 characters) to go out of bounds of the allocated input array.
-// The password is replaced with the out of bounds characters e.g. if `root--------123` is inputted for a username, the password is now `123`.
+/*
+    An example of a user changing the root password by attempting to login with a username that's
+    long enough (>10 characters) to go out of bounds of the allocated buffer array.
+    The password is replaced with the out of bounds characters e.g. if `root------123` is attempted
+    for a username, the password is now `123`.
 
-int checkInput(char input[], char toMatch[])
+    Requires an even number of arguments as 'username password' pairs. Attempts to login with each sequentially.
+    e.g. ./out-of-bounds-access.out root------123 bla root 123
+*/
+
+int checkCredentials(char *input, char *toMatch, char *successMessage, char *failureMessage)
 {
-    input[strcspn(input, "\n")] = '\0'; // strip \n from end of fgets
     if (strcmp(input, toMatch))
     {
-        printf("Wrong input\n");
+        printf("%s\n", failureMessage);
         return 0;
     }
     else
     {
-        printf("Correct input\n");
+        printf("%s\n", successMessage);
         return 1;
     }
 }
 
-int runAttempt(char input[], char username[], char password[])
+int runAttempt(char *username, char *password, char *buffer, char *usernameAttempt, char *passwordAttempt)
 {
-    printf("\nEnter username: \n");
-    fgets(input, 20, stdin);
-    if (!checkInput(input, username))
+    printf("\nChecking username: %s \n", usernameAttempt);
+    memcpy(buffer, usernameAttempt, strlen(usernameAttempt) + 1);
+
+    if (!checkCredentials(buffer, username, "Valid user", "Unknown user"))
     {
         return 0;
     }
 
-    printf("\nEnter password: \n");
-    fgets(input, 20, stdin);
-    return (checkInput(input, password));
+    printf("\nChecking password... \n");
+    memcpy(buffer, passwordAttempt, strlen(passwordAttempt) + 1);
+
+    return (checkCredentials(buffer, password, "Correct", "Wrong"));
 }
 
-int main(void)
+int main(int argc, char *argv[])
 {
+    if ((argc - 1) % 2 != 0) // argc counts program name as first argument
+    {
+        printf("Must be supplied an even number of arguments\n");
+        return 1;
+    }
+
     char username[] = "root";
     char password[] = "password";
```

Comment:
I tried making these global but then the memory allocation isn't adjacent with `buffer[10]`, which is needed for the out of bounds write to work.

Reply:
maybe you can make the pointers global but `malloc` them to ensure allocation order?

Reply:
done